### ISSUE TITLE
@textlint/browser

### ISSUE THREAD

**Issue by azu** [textlint/textlint#299]
We have noticed that some browser application use textlint.

- [textlint-app](https://github.com/textlint/textlint-app "textlint-app")
- [textlint.github.io](https://github.com/textlint/textlint.github.io "textlint.github.io")

Currently, [@textlint/kernel](https://github.com/textlint/textlint/tree/master/packages/textlint-kernel "@textlint/kernel") provide low-level API.
But, it is unusable and almost internal API. Because kernel's version < 1.0.

We want to provide high-level API for browser.

## Name

`@textlint/browser`

## API

Any idea?

- [ ] How to initilize?
- [ ] `lintText(text, extension): Promise`

## Limitation

All rules are not work on browser because some rules depended on Node.js

**Comment by azu** [comment]
I think that to support WebWorker is good thing.
[Web Workers to the rescue – Rastislav Svarba – Medium](https://medium.com/@ra100x/web-workers-to-the-rescue-4ecf4279953c "Web Workers to the rescue – Rastislav Svarba – Medium")

**Comment by Leko** [comment]
> How to load rule and plugin, and config?

I think it is better to define rules like a Babel.

```js
{
  rules: [
    [require('textlint-rule-prh'), {
      ruleContents: [
        `YAML content...`,
      ]
    }]
  ]
}
```

And using it as below:

```js
import { lintText, lintMarkdown } from '@textlint/browser'

const config = {
  rules: // like above format
}

const text = 'Hello'
const result = await lintText(text, config)

const markdown = '# Hello'
const result2 = await lintMarkdown(markdown, config)
```

> How to load .textlintrc?

I think we don't need support to load .textlintrc.
If anyone want to use .textlintrc, use with Webpack ([raw-loader](https://webpack.js.org/loaders/raw-loader/)):

```js
import { lintText } from '@textlint/browser'
import textlintrc from './.textlintrc'

const config = JSON.parse(textlintrc)
lintText('...', config)
```

@azu How about you?

**Comment by azu** [comment]
> I think it is better to define rules like a Babel.

textlint need to key of the rule/plugin.
This is the reason that [kernel](https://github.com/textlint/textlint/tree/master/packages/@textlint/kernel) require `ruleId` for opions.

```js
const options = {
    filePath: "/path/to/file.md",
    ext: ".md",
    plugins: [
        {
            pluginId: "markdown",
            plugin: require("@textlint/textlint-plugin-markdown")
        }
    ],
    rules: [
        {
            ruleId: "no-todo",
            rule: require("textlint-rule-no-todo")
        }
    ]
};
```

**Comment by Leko** [comment]
@azu Ah sorry.
Then, I think the better format is array of object.

```
    rules: [
        {
            ruleId: "no-todo",
            rule: require("textlint-rule-no-todo")
        },
        {
            ruleId: "rule-aaa",
            rule: require("textlint-rule-aaa"),
            option: { ... },
        },
    ]
```

> WebWorker support

I think we should support WebWorker.
Textlint does not depend on DOM and TextlintMessage is transferable.

**Comment by azu** [comment]
> I think we should support WebWorker.

Yes. Probaly, main use-case of textlint/browser is an editor like application.
That app run linting with high frequency.
This task should be split to non-main thread.

**Comment by azu** [comment]
Overview of textlint

![image](https://user-images.githubusercontent.com/19714/40547558-3e5d7122-606d-11e8-9f55-7100972fbd3c.png)

https://github.com/textlint/textlint/tree/master/packages/textlint/src

[`textlint`](https://github.com/textlint/textlint/tree/master/packages/textlint) includes CLI, Engine, and Core. (This term is historical reasons...)
[`@textlint/kernel`](https://github.com/textlint/textlint/tree/master/packages/%40textlint/kernel) is Kernel.

Probably, `@textlint/browser` is a sister/brother of Core.
Engine's tasks include loading of `.textlintrc`. 
As a result, we can drop loading task(load `.textlinrc` and parse it) from `@textlint/browser`.

**Comment by azu** [comment]
We might well have started with creating `@textlint/browser` as is a wrapper of kernel with WebWorker support.

>`@textlint/browser` = `@textlint/kernel` + WebWorker + α

@Leko How about you?

**Comment by Leko** [comment]
@azu I'm sorry for my late reply.

> Probably, @textlint/browser is a sister/brother of Core.

> @textlint/browser = @textlint/kernel + WebWorker + α

I agree with this plan.
I think core(`TextLintCore`) is unnecessary to lint in a browser.

**Comment by azu** [comment]
We can use [playground](https://github.com/textlint/playground) as MVP(Minimum Viable Product).
It is a just editor that work on browser.

I think that we should have experimental product(like playground app) which use `@textlint/browser`.
We can start to create a library after we can get actual needs.
It is good process to create good library.

Do you have any product that you want to create?

**Comment by Leko** [comment]
> We can start to create a library after we can get actual needs.

I think so too.

I already created markdown editor(uses monaco-editor, Textlint kernel and React Native for Web etc).
But it is for internal use. I can't publish it as OSS.

Another one, I got nothing for now.

**Comment by ylc395** [comment]
any progress? or anything we may help?

**Comment by azu** [comment]
I'm working on https://github.com/textlint/editor
textlint editor works on a browser.
